### DISCUSSION TITLE
Proposal: Loading translations nearby in pages/components of our websites

### DISCUSSION THREAD

**Discussion by joostdecock** [freesewing/freesewing#3418] [Forum]
## What is this?

This is a proposal for how we can load translations from *code-adjacent* (as in, not from a central `i18n` package of folders) for our websites that use translation, in other words freesewing.org and our development environment.

## What is this not?

A proposal to handle *code-adjacent* translation loading in patterns/packages. We'll get to that later.

## How it works

You can create a `namespace.en.yaml` file anywhere under these folders:

- `sites/org/pages`
- `sites/org/components`
- `sites/shared/components`

To be clear, you should replace `namespace` in `namespace.en.yaml` with the namespace you want to use.
So if you have a component called `Banana` it might make sense to create a namespace `banana`. So the file would be `banana.en.yaml`.

Crowdin is configure to keep track of these files, and will auto-create the `namespace.es.yaml`, `namespace.nl.yaml`, `namespace.fr.yaml`, and `namespace.de.yaml` files in the same location.

The `org` site's prebuild script will scan for all of these and auto-generate the namespace files where i18next hopes to find them.

## Using this in a page

Make sure to only load the namespaces you need in the page. This speeds up the initial page load.

Since you'll need to define the namespaces you want to use in more than one place, it's probably a good idea to define them at the top of the page. For example:

```js
// Translation namespaces used on this page
const namespaces = ['signup', 'errors']
```

**Note:** Do this **outside** the page component.

Then, you can load translations (inside the page component) like this:

```js
const { t } = useTranslation(namespaces)
```

And, **important** limit the loading in `getStaticProps` to only the namespaces we want:

```js
export async function getStaticProps({ locale }) {
  return {
    props: {
      ...(await serverSideTranslations(locale, namespaces)),
    },
}
```

## How does it work?

For the nitty-gritty, check this commit: https://github.com/freesewing/freesewing/pull/3421/commits/cb3231eba222da601177e22dc3ffa39c407047eb

**Comment by joostdecock** [comment]
Looks good on the crowdin side of things:

![Screenshot from 2023-01-10 20-44-24](https://user-images.githubusercontent.com/1708494/211647457-c9d4cedc-07a6-44ce-9d58-673e4fdd4238.png)

**Comment by eriese** [comment]
I think there's value add in even just putting an i18n folder at the root of each site and ([if I'm reading their very minimal docs correctly](https://github.com/i18next/next-i18next#4-declaring-namespace-dependencies)) using better namespacing to handle the reduction in bundle size. Like, right now because we don't know what namespaces we need, we're defaulting to 'app' in all components, and we could tighten that up. 

Where I'm struggling is here:
> For simplicity, we also drop a `i18n.mjs` file that holds this:
> 
> ```js
> import en from './en.yaml'
> import es from './es.yaml'
> import de from './de.yaml'
> import fr from './fr.yaml'
> import nl from './nl.yaml'
> 
> export const i18n = { en, es, de, fr, nl }
> ```

I wish there were a way to accomplish this without putting a duplicated file everywhere... How do we deal with maintenance when we add another language? Maybe generating this file can be part of the i18n prebuild process, like it scans the filesystem for localization files and adds this to any folders that have some?

> Now in your page or component, you can simply import that file:
> 
> ```js
> import { i18n } from './i18n.mjs'
> ```

I'm then using this like this, right? `const { t } = useTranslation('app', { i18n })`?;

> This is not the same as full translation support. For example, no support for interpolation. ( do{thing})

Why is that? because we're not importing all the strings?

**Comment by joostdecock** [comment]
> I think there's value add in even just putting an i18n folder at the root of each site and ([if I'm reading their very minimal docs correctly](https://github.com/i18next/next-i18next#4-declaring-namespace-dependencies)) using better namespacing to handle the reduction in bundle size. Like, right now because we don't know what namespaces we need, we're defaulting to 'app' in all components, and we could tighten that up.

I feel we might benefit from some more clarity of what's been happening in this realm already:

- 2021: Everything is in the i18n, the `app.yaml` file is a big mess of website strings
- 2022: Everything is still in the i18n folder, but the `app.yaml` file is no longer used for React components, instead [there's a new `components` folder](https://github.com/freesewing/freesewing/tree/develop/packages/i18n/src/locales/en/components) that has those strings. 
- 2023: We need to do something about translated strings

So when you voiced concern about the way translation strings were handled, I assumed it was because of they way we've been doing it recently. After reading your comments, I think it's more about how we used to do it traditionally.

With that in mind, one `i18n` folder per site is certainly an option. What I did here, is put translation as close as possible to the React components where they are used. The closest we can get is right inside the file itself, but crowdin does not support that.
So the next best thing is right next to it. That could be JSON, but JSON is a nightmare to maintain manually, so I opted for YAML and the relevant webpack loader.
Then we still need a way to load this into the component hence the `i18n.mjs` file. Yes, it's a bit silly to have several of these but abstraction has a cost too.

> I wish there were a way to accomplish this without putting a duplicated file everywhere... How do we deal with maintenance when we add another language? 

You are right of course. But can I point out how rare the event is you are describing. Ever since FreeSewing is multilingual, no language has ever been added. None. Ever.
So let's say this scenario you describe happens about once every 5 years or so.

I'm still open to an elegant solution to this, but I'm not sure it warrants extra complexity.

> I'm then using this like this, right? const { t } = useTranslation('app', { i18n })?;

No, the `i18n` in this example is an instance of react-i18next. 
To hook this into the translation system, we'd need to load the strings in `getStaticProps()` (see below)

> Why is that? because we're not importing all the strings?

This is a very simplistic example, and I am bypassing the translation system entirely here.
I'm just importing a bunch of strings in an object, and I can use them because we know what the current locale is.
These strings are (in this example) not yet loaded into i18next.

To hook this into the translation system, we need to load these strings in `getStaticProps()`. 
I need to look into the react-i18n API to check how to do this. By default it expects JSON files to be available at a specific path. That would require a prebuild step so I feel there's has to be a way we can just load them as props, and then it will be part of the regular Next.js build process.


I put together this example with a focus on:
- Where do we put the source strings for translation
- How do we get it to work with crowdin
- How do we load the translations into our pages & components

So I haven't spent any time on integration with i18next because that's not a matter of opinion, that's just a matter of getting it to work.

But these other things are much more a matter of opinion. So I guess that's what we're debating.

I acknowledge the use of the `i18n.mjs` file to tie it all together so to speak is a bit inelegant.
But it does allow keeping translation strings right next to the component using them.

And I feel that having one `i18n` folder per site is just going to grow into its own spaghetti of who even knows where these strings are used. Because in the end, `org` and `shared` will hold the vast majority of strings. 


> Maybe generating this file can be part of the i18n prebuild process, like it scans the filesystem for localization files and adds this to any folders that have some?

I think this is sort of the crux of the debate. Do we want to have translation files somewhere and then have a prebuild step that somehow puts everything in place? Or do we want to have it *just work* no prebuild required (like this example).

I'm not against embracing the prebuild approach. The main thing that I want to avoid is to just split our big mess of strings into two big messes of strings. 
One of the hardest things is knowing when it's ok to remove strings that are no longer used. So having them right in the folder of the component fixes that: Removing the component means removing the strings.

Let me try to refactor this a bit with a prebuild step, and see if we can keep the benefits of having the translations live next to the code without having to have *glue* files all over the place.

I'll also include integration with i18next to get a better picture of how this would work end-to-end.

Thanks for the feedback, good stuff 👍

**Comment by joostdecock** [comment]
@eriese (and others reading along), I've updated the proposal above after implementing some changes.

Long story short: 

- I've kept the idea of being able to put translation files anywhere (as in, right next tot he page/component using them).
- I've updated the i18n prebuild step to do what needs doing to make it *just work*
- The translation files now determine the namespace, and that namespace gets auto-loaded into i18next

I will also use this to further limit the translations we load on each page, because on many pages we just load everything which is often not needed. This should speed up the first paint.

Please let me know how you feel about this updated proposal.

**Reply by eriese** [reply]
an improvement, I think. Love to see an example of how this works in a component, though

**Reply by joostdecock** [reply]
Here's an example of this approach in a component: #3425

**Comment by eriese** [comment]
>So when you voiced concern about the way translation strings were handled, I assumed it was because of they way we've been doing it recently.

I will admit, most of my concern about where translation strings go was about design translation strings, which I'm hearing are outside of the scope of this proposal for legitimate reasons. With site translation keys, I agree we want it cleaner and we're already making progress on that front so I'm not sure an entire overhaul is the priority. Possibly just better guidelines on where things go, how they're named, and how to keep your scope small. 

> And I feel that having one i18n folder per site is just going to grow into its own spaghetti of who even knows where these strings are used. Because in the end, org and shared will hold the vast majority of strings.
One of the hardest things is knowing when it's ok to remove strings that are no longer used. So having them right in the folder of the component fixes that: Removing the component means removing the strings.

I don't know the specifics of whether this is really true or not, but my sense is that getting translations out of their own package and into the sites will make it easier for us to use [extraction tools](https://www.i18next.com/how-to/extracting-translations) or [something like this one that scans for unused keys and missed key](https://github.com/mxmvshnvsk/i18n-unused) to remove some of the overhead of figuring out what we can delete. We may still need something centralized for truly universally needed strings like "yes" and "no"... ugh

> This is a very simplistic example, and I am bypassing the translation system entirely here.
I'm just importing a bunch of strings in an object, and I can use them because we know what the current locale is.
These strings are (in this example) not yet loaded into i18next.
I'll also include integration with i18next to get a better picture of how this would work end-to-end.

I definitely wouldn't want to sacrifice the "it just works" of using `i18next` to keep the strings closer to their components. And while I think that for the most part a person who is making a whole new component will be someone who can look at an existing component to see all that goes into it, it's feeling like a lot of additional steps are needed to import the right strings, set them up, and access them properly. With any luck you find that that the i18next integration is as trivial as you're imagining, but I'm not holding my breath.

> I think this is sort of the crux of the debate. Do we want to have translation files somewhere and then have a prebuild step that somehow puts everything in place? Or do we want to have it just work no prebuild required (like this example).

I love the idea of fewer or no prebuild steps; I think that would be wonderful. I just don't want to have to grow our own i18n library to achieve that. Also to keep in mind is the standalone environment and how it will receive translations in this new plan.

**Reply by joostdecock** [reply]
Alright, so I should probably clarity some of my thinking on this (which has evolved):

**we're not building a i18n library**
The solution implemented now it entirely based on next-i18next (which uses i18next) under the hood.
Because indeed, we don't want to start building our own library for this.

**we need a prebuild step**
It would be nice if it could *just work* but to get close to that we'd have to lean heavily on webpack. And there's no webpack when we are packaging up designs (since we probably want to also keep translations of designs/plugins next to them).

So let's just swallow our pride and extend the already existing i18n prebuild step to gather all the translation strings from all over the place and package then neatly in the way that i18next expects them, and voila, things just work.

**available namespaces**
Once thing that's tantalizing is to optimize the bundle size by only loading those namespaces that we need.
This needs to happen in `getStaticProps()`, so at the page level.

For some pages, that will be relatively straightforward (like the signup page for example).
For others (like the single page that is used to generate all documentation) this is not so obvious as the names of options (for example) of all designs are used in the navigation. But on the other hand, we don't need the strings used on the account pages there, and so on.

Note that currently we include all namespaces. So this is not a new issue, but rather an area where we can potentially speed up things.
To make that somewhat manageable, I propose that React components should export a named export called `namespaces` that lists the namespaces the component relies on.
Or at least, I plan to road-test that idea for freesewing.org

I've setup a PR with an example of this approach here: https://github.com/freesewing/freesewing/pull/3425

**Reply by eriese** [reply]
> Note that currently we include all namespaces. So this is not a new issue, but rather an area where we can potentially speed up things.
> To make that somewhat manageable, I propose that React components should export a named export called `namespaces` that lists the namespaces the component relies on.

So each component will import the namespaces its children use, concat those with its own namespaces and export that up? That feels workable if we must, but also kind of fragile

Are we opposed to a little bit of lazy loading? https://github.com/i18next/i18next-http-backend/tree/master/example/next

**Reply by joostdecock** [reply]
> So each component will import the namespaces its children use, concat those with its own namespaces and export that up? That feels workable if we must, but also kind of fragile

We must not :) We can just keep loading all namespaces for all pages, as we've been doing all along.
But I would like to see if I can do this for the homepage where:

- We only need a limited number of namespaces
- Always nice to make the homepage load faster as it will be the entry point for many people

So think of this more as something I'd like to try in specific scenarios, rather than something I'm expecting everyone to do all the time.

> Are we opposed to a little bit of lazy loading?

I'm not opposed to lazy loading, but my main goal here is to try to limit the amount of javascript/json that needs to be parsed before bootstrapping the SPA.
That initial load is the SSR (server-side render) code. So this solution would not impact that, as it's client-side only.

That being said, I do think we should look into lazy loading for the pattern drafting UI.

**Comment by joostdecock** [comment]
Done. Closing
